File: sections/30_gcd_lcm_diophantine/problems/c_linear_diophantine/solution.cpp
#include <bits/stdc++.h>
using namespace std;

long long egcd(long long a, long long b, long long &x, long long &y) {
    if (b == 0) {
        x = 1;
        y = 0;
        return a;
    }
    long long x1, y1;
    long long g = egcd(b, a % b, x1, y1);
    x = y1;
    y = x1 - (a / b) * y1;
    return g;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int q;
    if (!(cin >> q)) return 0;
    while (q--) {
        long long a, b, c;
        cin >> a >> b >> c;
        long long x, y;
        long long g = egcd(a, b, x, y);
        if (c % g != 0) {
            cout << "IMPOSSIBLE\n";
        } else {
            long long scale = c / g;
            cout << x * scale << ' ' << y * scale << '\n';
        }
    }
    return 0;
}

File: sections/30_gcd_lcm_diophantine/problems/c_linear_diophantine/solution.py
import sys


def egcd(a: int, b: int) -> tuple[int, int, int]:
    if b == 0:
        return a, 1, 0
    g, x1, y1 = egcd(b, a % b)
    return g, y1, x1 - (a // b) * y1


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    q = data[0]
    idx = 1
    out = []
    for _ in range(q):
        a, b, c = data[idx], data[idx + 1], data[idx + 2]
        idx += 3
        g, x, y = egcd(a, b)
        if c % g != 0:
            out.append("IMPOSSIBLE")
        else:
            scale = c // g
            out.append(f"{x * scale} {y * scale}")
    print("\n".join(out))


if __name__ == "__main__":
    main()
